### ISSUE TITLE
Provide access to compile/parse errors generated during wrenInterpret()

### ISSUE THREAD

**Issue by preble** [munificent/wren#265]
Wren presently outputs parse errors to stderr. I would like to see an API to access the text of these errors programmatically, either by querying for them after a failed `wrenInterpret()`, or (better?) via callback supplied in `WrenConfiguration`. If the user is able to supply their own Wren code (as I would like them to be able to do) we will need a way to show them the errors.

The callback approach might work as a way to report errors during `wrenCall()` as well?

**Comment by munificent** [comment]
Yes! This has always been my intention, I just haven't gotten to it. A callback in `WrenConfiguration` is exactly what I had in mind.

If you'd like to take a stab at implementing that, I'd be happy to guide you through it. I'm not sure if the varargs stuff will make it harder, but we should be able to figure it out.

**Comment by preble** [comment]
I'd be happy to take a shot at it, thanks!

Would this callback be specifically for `wrenInterpret()`, or should it be more general – for runtime errors from `wrenCall()`, too? Only difference I can think of at the moment is that it'd be great for `wrenInterpret()`'s callback to provide line/column information via a struct. With `wrenCall()` that kind of information isn't applicable.

**Comment by munificent** [comment]
> With wrenCall() that kind of information isn't applicable.

Right. On the other hand, callstack information is.

How about we start with one and then expand it to handle both compile and runtime errors over time? It may be that we end up with two separate callbacks, one for compile errors and one for runtime. If possible, though, it's good to try to keep it simple.

**Comment by preble** [comment]
The varargs/formatting bit turned into a bit of a rabbit hole. The approach I was starting to take was to build out a string structure (like `String` but with capacity too), and then functions to allocate, free, and vsnprintf to it (reallocating as needed). All of this in the interests of being able to pass the formatted null-terminated C string to the callback. 

As I type it out here it doesn't sound that bad, but when I was writing it it felt overengineered. Thoughts? Is there a simpler way that I'm not thinking of? I'm trying to steer clear of anything unsafe or too constrained (like limiting output length).

**Comment by munificent** [comment]
> The varargs/formatting bit turned into a bit of a rabbit hole.

Yeah, that's part of why I kicked it down the road so far.

> As I type it out here it doesn't sound that bad, but when I was writing it it felt overengineered.

It would definitely be nice to not have to reimplement much of stdlib streams inside Wren. After some digging around, it turns out there is a simple solution. Woo!

Alas, it's C99-specific, so a little hackery will be required to please MSVC.

The solution is to use `snprintf()`, *twice*. It is specified to return the number of bytes it *would* have written if the buffer was large enough. So you can call that once to tell you how big of a buffer to allocate. Allocate that. Then call it again to fill in the buffer.

[This thread](https://social.msdn.microsoft.com/Forums/vstudio/en-US/2b339bdf-7ab1-4a08-bf7e-e9293801455b/c99-standard-snprintf-function-?forum=vcgeneral) has some details about how to get it working in MSVC too. Does that give you enough to go on?

**Comment by ppvk** [comment]
@preble, did your attempt look kind of like #333 ? I'm just handing a function the same content `fprintf` would normally consume.

**Comment by preble** [comment]
@ppvk I regret that I completely dropped the ball on this. I'll have a look at #333, though!

**Comment by preble** [comment]
@ppvk To be honest I didn't end up attempting it – life got rather busy. :frowning: Looking over your PR, I see some similarities to what I had imagined – especially the `errorFn` – although I had a few different ideas, specifically adding a centralized `wrenPrintError` method to make it easier to write strings to it and centralize that logic. There is also the issue of MSVC compatibility to contend with. I ended up getting into it enough that I took a pass at it; I'll put up a PR in a bit. I'd welcome your feedback as well!
